### ISSUE TITLE
[architecture] modules per collection and/or doc

### ISSUE THREAD

**Issue by mesqueeb** [vuex-cloud/rfcs#10]
TLDR;
- we'll have resource (collections) modules and sub-resource (documents, sub-collections) nested modules
- "collection" modules can receive the _read_ actions and the _insert_ action
- _read_ actions triggered on the "collection" results in the retrieved documents to be dynamically registered as sub-module
  - the dev can provide a template for this
- "document" modules can have _read_ and _write_ actions
- both type of modules are able to have custom actions/mutations/getters/state
- it's recommended not to place custom state values on a  "collection" module since this is also where all docs will be added
- the retrieved data of a document will by default be placed into a sub-prop (`_`) on the state of the "document" module. (this can be configured to be another name or `null` to have the data added to the root of the state)
- each module will have a "metadata" prop (name tbd) with information on things like the timestamp docs were fetched last etc.

--------------------

This thread will be to discuss:

#### Current implementation in v-e-f:

The dev chooses if a module is a **'collection' or 'doc' module**. Usage of the API slightly changes based on this. See the [docs](https://mesqueeb.github.io/vuex-easy-firestore/add-and-manage-data.html#basic-usage) for a refresher on v-e-f's differences of the API usage with 'doc' and 'collection' mode.

#### @louisameline 
> There is currently just one module for a collection, and its document are added in its data. Tomorrow this should change, we'll have one module for the collection which will instantiate as many modules per document as needed. So the default values of the document will sit in the state of the document module, and `DefaultValues` can go away. We'll just need a way to tell the collection module which module to use for its children.

> With submodules, a collection module will declare document submodules. You would call the delete action on the submodule only. It's actually weird to call a parent and ask it to delete a prop on its child if you think about it

> Side note: I think we need to think less in collection/document model. Let's think rather in resources. A resource can be the child of another resource. I would get rid of the "collection" concept all together, for example no "getCollection" or that kind of things. You just get() a resource or getChildren()". Document/Collection probably won't apply to other databases we'd like to plug onto vuex-cloud in the future.

> `dispatch('myCollection/get', '123abc')` assumes that we have to call to the collection module to retrieve a document, it's a pain. Isn't there a way through vuex plugins to listen to request, and to instantiate a module on the fly, so we could call `dispatch('myCollection/123abc/get')` even when the module has never been instantiated by the user or a parent collection?

#### @mesqueeb 

No, in Vuex it's not possible to call actions from a module that doesn't exist.
`dispatch('myCollection/123abc/get')` will never be possible, neither will creating docs like so: `dispatch('myCollection/123abc/insert')`.

The only way that would be possible is if we use our own custom modules via the new vue reactivity api, because they we're not bound by the "dispatch" limitations anymore. It's just functions.

Allow me to first give an example with the regular vuex method where we'll have to be forced to use a "collection" module to create and read/retrieve data.

### Implementation 1: collection module + dynamic sub modules

Allow me to use a Pokédex "collection" for the sake of examples.

```js
// inserts are done on the collection:
dispatch('pokedex/insert', {name: bulbasaur, level: 1})

// read actions as well:

// get all pokemon from the pokedex
dispatch('pokedex/get') // gets everything
// or
dispatch('pokedex/getById', '001')
// or
dispatch('pokedex/stream', {where: [['type', '==', 'water']]})
```

each these "read" actions will dynamically register a module for the document, but we'll still need to use the "collection" module to be able to get all the docs etc.

```js
// show all the docs:
Object.values(store.state.pokedex)
```

If the dev needs a getter for a specific prop, he'll have to provide the template of the module that'll be used each time a doc-module is registered dynamically.
So he'll have to make some sort of config file for this.

We can also have the user register a sub module called `new` from the start which can be used to create docs and provide the template getters for the retrieved docs:

#### setup

```javascript
// pokedexCollection.js
import { insert } from '@vuex-cloud/setters'
import { get, stream } from '@vuex-cloud/setters'
import pokedexDocTemplate from './pokedexDocTemplate.js'

export default {
  // collection module
  state: {},
  actions: { insert, get, stream },
  // register the doc sub-module called 'new'
  modules: {
    new: pokedexDocTemplate
  }
}
```
```js
// pokedexDocTemplate.js
import { update, delete } from '@vuex-cloud/setters'

export default {
  // doc module
  state: {},
  actions: { update, delete },
}
```
```
// store.js
import pokedex from './pokedexCollection.js'

export default {
  modules: {
    pokedex
  }
}
```

#### usage

```js
const id = await dispatch('pokedex/insert', newDoc)

// then the new doc can be read like so:
store.state.pokedex[id]
// or edited
dispatch(`pokedex/${id}/update`, newData)
```

Please note: This requires the dev to manually add a sub-module called `'new'` or `'template'` or whatever with the actions and getters he wants for each dynamically registered sub-module.

#### merits

- The dev could add his own actions / getters to the sub module called "new" and this could be used as the template for each document module that's registered dynamically.

#### demerits

- This method means we won't be able to put the docs retrieved from the server into a `data` prop, because the state of each sub-module is attached directly to the collection state.

**Comment by mesqueeb** [comment]
Sorry I ran out of time today. The above is still a WIP. 
I'll continue tomorrow. ;)

**Comment by louisameline** [comment]
Can't wait to talk about this :)

**Comment by mesqueeb** [comment]
@louisameline I have further thought about this and have updated my post above.
I realised we cannot use the `data` subprop to save docs, because sub-modules save their state to the top level of the parent module.

I don't think there's a better way than Implementation 1...

I thought about a custom class for collections.
We'll be free in what kind of functions we use for inserting and reading data with this. But after writing up some ideas, I realise this is way too convoluted and confusing.

You don't even need to read this:

<details>
  <summary>Implementation example 2 (custom class for "collections")</summary>

### Implementation example 2

We can use a custom class to represent collection modules.

```javascript
// this is a simplified representation of what Collection could look like:
class Collection {
  constructor (collectionName, collectionModule = {}, preventDynamicRegister) {
    this.collectionName = collectionName
    this.collectionModule = collectionModule
    const path = collectionName
    // register the "collection" module dynamically or not (see setup)
    if (preventDynamicRegister) return
    registerDynamicModule(path, collectionModule)
  }
  
  // instead of the collection "module"
  // this class has the insert method:
  insert (data = {}) {
    const id = data.id || randomId()
    const path = `${collectionName}/${id}`
    // register the "doc" module
    registerDynamicModule(path, data)
    syncToServer(collectionName, id, data)
  }

  // the read actions just register dynamic modules:
  stream (filters = []) {
    const docs = streamDocs()
    docs.forEach(doc => this.insert(doc))
  },
  get (filters = []) { /* ... */ },
  getById (id) { /* ... */ },
}
```

#### usage

Throughout the app it could be used dynamically:

```javascript
const pokedex = new Collection('pokedex')
pokedex.stream()
pokedex.getById('001') // bulbasaur!

pokedex.insert({name: 'squirtle'})
```

#### setup

It could also be used in the initial setup like so:

```javascript
import { Collection } from '@vuex-cloud/setters'

// prevent dynamic registration because there is no store yet
const pokedex = new Collection('pokedex', {}, false)

const store = {
  modules: {
    pokedex: [pokedex.collectionModule]
  }
}
```


</details>

I have also thought of using a custom class for the sub-modules, but really, I think we should just stick to pure Vuex modules right now.

**Comment by louisameline** [comment]
Ok, I'll be thinking out loud here, I don't have all the answers right now.

> I realised we cannot use the data subprop to save docs, because sub-modules save their state to the top level of the parent module.

Right, I checked, the parent module does have the state of its children modules's under a prop of their name. So a collection module that registers a doc as a nested module will have a `state.docId == { ...docState }`. The `state` of the child however, when access from an action within the child itself, is only its own as expected.

Could we list the reasons why this would be an issue though?
- If the user stores his properties at the root of the module, they will be side by side with the states of submodules: it's ugly and can conflict.
```javascript
state: {
  myModuleProp: true,
  myModuleProp: {},
  metadata: {},
  mySubmoduleName: {}
}
```
So it's clear we have force the use of the `data` subprop:
```javascript
state: {
  data: {
    myModuleProp: true,
    myModuleProp: {},
  },
  metadata: {},
  mySubmoduleName: {}
}
```
- There is an issue if the name of a submodule is `data` or `metadata`. We could name them `resData` (res for resource) and `resMetadata` or whatever to prevent conflicts.
- Not very elegant for the user to write `$store.state.mycollection.documentId.resData.docProp` in his code...

However the deeper issue here might be Vuex's way of accessing properties through a path.
As a user, how do I know if `$store.state.a.b` is a property of `a` or a resouce on its own?
I'm not an expert in network theory, but with REST APIs, we are supposed to used the path to link to the resource (which has its GET/POST/etc. methods), not to the property of the resource, right? So Vuex's `$store.state.module.propname` accessor system cannot (when used with its default paradigm, see below) reflect locally a RESTful remote API. It seems to me that Vuex's submodules were an afterthought.

We're not screwed however, it just means that we shouldn't use the `$store.state.module.propname` getter system. Vuex should probably have offered a `$store.state.module.read('propName')` method, which would have returned the own property of the module.
We could provide such a function by adding a `read` function at the root of the state of each module. However, if the user has a submodule named `read`, it will get overwritten.
So we could use a getter: `$store.getters[moduleName/read'](propName)`
It works and is conform to the Vuex way of doing things, but it's cumbersome. So additionally, I would also offer this nice shortcut `$store.read('moduleName', 'propName')`.
So, vuex-setters would include a getter after all.

About the class you wrote: I think you did that basically to mitigate the fact that we currently can't write `dispatch('collection/doc/insert', payload)` when the doc module isn't instantiated, right?

- My first thought is that we should think in generic terms of resources, not collection/document.
- My second thought is that yes, your class gets too far from Vuex, it's no longer "Vuex built-in cloud support".
- My third thought is that we might wrap the dispatch method: `expedite('collection/doc/insert', payload)` => it would check if the module exist and register it if not, and then dispatch the `insert` action. But there are other questions that need to be investigated here to know if it's feasible without having to read Vuex internal properties.
- My final thought: should we fork Vuex to address the above issues properly? Or get in touch with its authors to know if they would do something about it?

**Comment by louisameline** [comment]
In this post, Linus Borg is like "Vuex is built for simple things, not reproduce a database": https://forum.vuejs.org/t/totally-confused-by-vuex-and-a-rest-api/32977/2

However this position is not realistic IMO. It's at the core of the Vue system, you can't promote it so much and then, when the app gets a little more complex, say "oh but it's just a toy you know, find something else if you have real data to organize and a backend to work with".
I haven't found any one else complaining about the way submodules work, but I think we have decent arguments to open an issue.

**Comment by mesqueeb** [comment]
@louisameline 

I'm not 100% sure about what you mean with everything you said in the post that starts with:
> Ok, I'll be thinking out loud here...

In my head I kinda already took peace with creating the next version with the choice of modules being a "collection" or "document" module, just like in v-e-f.

- A "document" module could have read & write actions and has no notion of IDs, only props.
- A "collection" module can only have read actions & the "insert" action. Other actions like "update" and "delete" must be presented on the dynamically registered sub-module which will automatically be a "document" module.

This is what I explained in my last post as well, and I hope we're not misunderstanding anything related to this.

### Some small comments about the parts I _did_ understand about your post

>> I realised we cannot use the data subprop to save docs, because sub-modules save their state to the top level of the parent module.

> Could we list the reasons why this would be an issue though?

Actually, not a big problem. The only "downside" is that, if the dev decides to add his own props to the "collection" module's state, he would need to filter them out when doing things like `Object.values(state.myCollection)` to get all documents. → However I think having the dev taking care of this on his own (he can make getters that filter out the props he added hisself) is best.

So never mind about this point! 😅  But we really need to chat, because I don't understand what you are trying to say with the rest of your post.

### Moving forward with my current understanding

With this in mind - and only IF we decide to go forward with this in mind - we should move the conversation about _how_ we will implement some core functionality like events & triggering sync actions after performing the local actions of the setters etc.

For the sake of our conversation going on here, I'm just gonna write some thoughts concerning this down below:

1. You say to provide configuration on the top level of the module (not in the state)

I did thought of this, but in the case we make the dev _just add actions he wants_, those actions have no ability to ever read that config. Because the scope inside an action has no access to the module's top level.

My point:
```js
import { update } from 'setters'

export default {
    vuexCloudConfig: {/* ... */},
    state: {},
    actions: {
        update, // ← this action can't ever access `vuexCloudConfig`
    },
}
```

Therefor I realised, we can just create functions that "create" actions and feed them the config object. Example:

```js
import { createUpdate } from 'setters'
const vuexCloudConfig = {/* ... */}

export default {
    state: {},
    actions: {
        update: createUpdate(vuexCloudConfig),
    },
}
```

If this "vuexCloudConfig" is passed to each setter imported from vuex-cloud and also to the "read" actions from "@vuex-cloud/firestore", then every action that is triggered will understand the entire configuration of the vuex-cloud enhanced module.

This allows things like setting up hooks like so:

```js
const config = {
  firestorePath: 'coll/doc',
  on: {
    insert (
      {state, dispatch},
      payload,
      updateFn
    ) {
      return updateFn(payload)
    },
  }
}
```
... this is just a hypothetical example of how eg. a hook could be set up.

Also, initially when `createUpdate` is executed which returns the `update` setter, the config could be examined for things like `firestorePath`. When a `firestorePath` is found, the action `update` would then execute the sync and return the sync promise instead of only updating the local store.

However, this design would be similar to the current v-e-f, and I'm having trouble figuring out how it's possible to extract all sync functionality into a separate `vuex-cloud/firestore` package with such an architecture.

So together with the meaning of your previous post, this is exactly what I'd still wanna discuss with you to see what your ideas are.

**Comment by louisameline** [comment]
Before we talk about the setup part, following up on our discussion about internal struture of data:
the getter solves another issue that I already wanted to tackle: retrieving the data

**Comment by louisameline** [comment]
Before we talk about the setup part, to follow up on our discussion about internal struture of data, I have more considerations:

**A.** on the bright side, the `$store.read('collection/syntax', 'propName')` syntax can be nice as it could (with parameters as 3rd option maybe) transparently instantiate the doc module at collection level under the hood, retrieve the doc and serve the value
**B.** on the down side... Vuex's structure is not great to iterate on all docs of a collection in the template, as `v-for="child in $store.state.myModule"` will stumble upon the `data` and `metadata` properties. We could think of another accessor but...
**C.** I realized that `$store.read('moduleName', 'propName')` would not be reactive in a template, so we need to find another solution. I see 2 ways to go about this:

**1. We keep thinking very close to Vuex**

Let's say we want to keep on working with the `$store.state.myModule.prop` syntax.
- To mitigate B, we could create a reference from `$store.state.myModule.children.childId` to `$store.state.myModule.childId` so that the user could do `v-for="doc in $store.state.myCollection.children"` without issues. `children` would have to be a non-enumerable property to avoid being serialized and saved by vuex-persist.
- To mitigate C, we forget about the getter and let the user work with `$store.state.myModule.data.prop`. We could we rename `data` by `_` if it looks better, and explain why this convention is needed to avoid conflicts. So it would be like `$store.state.myCollection.myDocument._.docPropName`. Not great but we don't really have a choice.

**2. Thinking "around" Vuex**

Since `$store.state.myModule` accessor syntax is fundamentally flawed for our usage, we need something else.
I've thought of mappers but came up with something better I think. What if we provided another `state` object? Something like `this.$$state = VuexCloud(this.$store.state)` and then we could call this in the template:
```javascript
// resolves to $store.state.collectionId.docId.data.docPropName
$$state['collectionId/docId'].docPropName
// we create a `children` non-enumerable property that allows iterating on children resources
v-for="document in $$state['collectionId'].children"
```

So basically this "flattens" the real Vuex store, preserves reactivity and still keeps the props of each resource nicely isolated. For the rest, Vuex can be used normally (getters, actions). What do you think?

*Note: the `children` property would generate a conflict if a resource had a prop of the same, so it should be called something less frequent like `resChildren` or something*

**Comment by louisameline** [comment]
This last solution I just gave could be useful for anyone who uses nested modules in Vuex, not just people who use Firestore. This could live in its own repo too: `vuex-deep`.

To support `vuex-deep`, `vuex-setters` will only need to have a `prefix` option which will be `'data'` (or `resData` or whatever key we use to isolate a module's data).

I don't know very precisely yet how we'd write `vuex-deep` when it comes to reactivity (Vue 2 support or Vue 3?) but it sounds like fun, I'm pretty hyped right now hehe

**Comment by louisameline** [comment]
I have this feeling that in addition to the standalone `vuex-setters` and `vuex-deep`, we could do a full package `vuex + vuex-setters + vuex-deep = vuex2` (or whatever the name). It's vuex wrapped up out-of-the-box with our new set & get syntaxes, that make up for the lacks of vuex. Man this is exciting, let's talk tomorrow.

**Comment by mesqueeb** [comment]
## 1. We keep thinking very close to Vuex

I think, if we want to attract people who already use vanilla Vuex specifically,
we should: **1. We keep thinking very close to Vuex**

## 2. Thinking "around" Vuex

If we decide to go this route, even _with_ these demerits, I'd rather do it **in a very different way.**

Take for example first, the trend that's already gaining _some_ traction: [vuex-module-decorators](https://championswimmer.in/vuex-module-decorators/pages/overview.html#what-it-does).

### What they try to solve (which I agree with since 2017):
- no type safety for parameters passed to commit & dispatch
- no autocompletion when typing these mutation and action names

### Their solution:
- a typescript module setup that **compiles to regular vuex**, but also provides wrappers that give the above benefits.

### My idea:
Even before I saw this package I already had played around with the composition-api for Vue 3.
They exposed the reactivity so even now, everyone can use it (even in Vue 2 with a plugin).

And I showed this before but, [this post](https://github.com/vuejs/rfcs/issues/72#issuecomment-514718718) just made so much sense to me, and is the same exact thing I have been thinking about for so long:

<img width="698" alt="image" src="https://user-images.githubusercontent.com/3253920/69908280-6477aa00-1429-11ea-8c55-a17b7a964fd2.png">

The vue-composition api (exposed reactivity) is available as [plugin for vue2 here](https://github.com/vuejs/composition-api). And its syntax is the same as vue3's syntax.

I'm just saying, if we're not going to stay close to vanilla vuex, we will get so much more benefits from using this way of creating modules.

### Benefits:

- We're free to configure our state & module structure how we want
- We're not bound to the bad vuex syntax (the 2 problems above)
- We're not bound to limiting our actions to only have 1 single parameter
- We'll be streamlined to the syntax of Vue 3 before any other library - this could be used as marketing point (Vue 3 plans to release in about 4 months)

### Demerits:

- Our name is `vuex-cloud`... But I think, even without using vuex, we can still use `vuex-cloud` if we state that `vuex` just stands for "central state management" in general. 😄 

---

TLDR;
- If we choose # 1 above:
  Let's stay as close as possible to the current way Vuex works - let's just make it work with what we have.

- If we choose # 2 above:
  Let's use [the Vue composition-api plugin](https://github.com/vuejs/composition-api) and make our modules work _just_ like Vue components without templates. (many many merits, few demerits imo)

**Comment by louisameline** [comment]
Interesting. I think we can both agree to stick close to Vuex *for now*. We'll be free to go one step further in either direction later.

I am not in favor of a Typescript solution, because I think it will deter so many people from the library. I hope Typescript itself will be replaced at some point by something better, but that's a personal opinion.

I think that while `vuex-deep` is only a small step on the side from Vuex (only the way data is structured in the exposed state actually changes), or even rather an improvement of Vuex, your proposition is actually something very different, and Vuex  in this context seems almost like a legacy dependancy that we'd want to get rid of. Although it's without any doubt a totally legitimate solution, it kinda puts me off. It's a nice thing to have a single "standard" store solution in the Vue community, I want to build on top of it and not dig its grave. For all I know, Vuex might  be already dead but still, for now I wish to stick to it, but could totally reconsider in the future when we see how things goes.

**Comment by mesqueeb** [comment]
@louisameline 
How about we start with the solution without your idea of `vuex-deep`.
Then once we the core is there you can work on `vuex-deep` and I can focus on other parts?

Since your idea of `vuex-deep` is additive, so I don't see it stopping us from starting on the core.
What do you think?

I don't wanna write off vuex-deep or anything 😉 

(PS why not call it vuex-flatten-state, or something?)

**Comment by louisameline** [comment]
Yes let's go without it for now, that's what I was suggesting as well :)
Same reason `vuex-setters` should not be called `vuex-actions` actually: flattening the state is not the purpose of vuex-deep, the purpose it to be able to work properly with data structures which have several levels (nested modules), hence the "deep" name.

**Comment by louisameline** [comment]
So, what should we call the property wrappers? As a reminder, the names must not be common to avoid conflicts with user's module properties. Do you like the `_` for the user's data, `resMetadata` for the lib internals, `resChildren` to iterate on sub-resources? Other ideas?

The underscore, as in `$store.state.myCollection.myDocument._.docPropName`, is debatable, because it makes us think of "module private data":
- it's data which belongs to the module itself indeed, not to a submodule
- on the other hand the data is actually not private, it's used by the dev and synchronized to the backend

But since the underscore here is not actually the first character of a full propName, I think it's ok.
It's short and non-invasive, I'd personnaly go for it. Let me know your thoughts

**Comment by mesqueeb** [comment]
I kinda like the props `data` and `meta` myself. I checked my v-e-f apps, and realised I had some with "doc" modules with the `statePropName` set to `data` as well. This means that the term `data` for me is already used heavily in the templates, and will make the upgrade very easy.

Others might have conflict with `data`, but others might think `_` is _too_ abstract...

But it's really not difficult to implement a choice for the dev (like `statePropName` in v-e-f). I think we can just give the choice to the dev, then everyone is happy.

**Comment by mesqueeb** [comment]
Allow me to try and wrap this thread up:

TLDR;
- we'll have "collection" and "document" modules
- "collection" modules can receive the _read_ actions and the _insert_ action
- _read_ actions triggered on the "collection" results in the retrieved documents to be dynamically registered as sub-module
  - the dev can provide a template for this
- "document" modules can have _read_ and _write_ actions
- both type of modules are able to have custom actions/mutations/getters/state
- it's recommended not to place custom state values on a  "collection" module since this is also where all docs will be added
- the retrieved data of a document will by default be placed into a sub-prop (`_`) on the state of the "document" module. (this can be configured to be another name or `null` to have the data added to the root of the state)
- each module will have a "metadata" prop (name tbd) with information on things like the timestamp docs were fetched last etc.

please go ahead and edit my post and close the thread. 😄 

next we should continue discussing the other issues in the first column:
https://github.com/vuex-cloud/rfcs/projects/1

**Comment by mesqueeb** [comment]
Hey @louisameline I think we can close issues we agreed on, so we can check the [roadmap](https://github.com/vuex-cloud/rfcs/projects/1) for the topics left. 😉

**Comment by louisameline** [comment]
We'll have to settle on the details though, but ok!

**Comment by louisameline** [comment]
Hi Lucas! I've been working on my app again (I'm so late on my schedule), sorry for the delay of response on our matters. A couple considerations to keep in mind:
- When a user unregisters, I have to destroy his data at Firestore. Most of the time, I handle the data with document modules, and notice there was no `delete` method on them (only their properties can be deleted, not the document itself). I understand the reasons you didn't do it, but we should make it possible in vue-sync though. I don't want to create a collection module only to be able to delete one doc. I dirty-patched my version of the library until vue-sync is out, so that a `delete` call without argument deletes the document and resets the module state.

- It opens the discussion on how it can be achieved in vue-sync. The issue being that modules which are initially (non-dynamically) declared in Vuex cannot be unregistered. So, 2 possibilities:
  - we declare *all* modules dynamically ourselves somehow, so we can unregister any one of them
  - we let the user declare his modules, but we need a way to mark them as "deleted", and throw errors when the user calls actions on it

The second option is not great because the module isn't really destroyed and it might be tricky to prevent all interactions with it. The first option sounds much cleaner, although I do not have all the implications in mind.

**Comment by mesqueeb** [comment]
Hi Louis.
### About deleting a single doc in vef 1.0
With vef-1.0 my original intention was to provide "doc" modules for docs that always have the same path, like "user settings" at a location like `users/{userId}/userData/settings` or something.

If you need to delete the actual doc called "settings" it'd be possible by having a "collection" module for `users/{userId}/userData` and deleting the `settings` doc.

I see now that in your use case - unregistering a user - this is not an elegant solution at all.

### About deleting a single doc in 2.0
We had already agreed that the "delete" function should be on the "doc" module, so this shouldn't be a problem for 2.0 I believe.

### About transactions
I think we can provide this option as well in the "firestore" plugin, but I think we'll have to think about that when we get our "core" lib running and we start developing the plugins. I don't have a clear vision about it atm. 😅 

### About dynamically registering modules
Originally I was thinking of having "collection" modules set up on beforehand by the dev, and "doc" modules dynamically being registered as sub-modules of those collections.

However, I believe **it'd be good if we have everything dynamically registered**, even collections, so people can query entire new sections mid-lifecycle. Thanks for this great idea!!

Let's make sure this is possible right from our core architecture! ;) I don't think it'll be too hard.

**Comment by louisameline** [comment]
Great!
The transactions could have an impact on how we do things in the core. I laid a first couple thoughts in a dedicated issue #17

**Comment by louisameline** [comment]
As discussed today, some thoughts on the structure of the local store

```javascript
// a clean structure
this.$store.state = {
  collectionModuleA: {
    _: {
      metadata: {},
      local: {}
    },
    // submodules are here because of Vuex works like that.
    documentModuleB: {},
    documentModuleC: {}
  },
  documentModuleD: {
    _: {
      data: {},
      metadata: {},
      local: {}
    },
    collectionModuleD: {},
    collectionModuleE: {}
  }
}

// the `_.data` is bad as it unnecessarily exposes the internals of our system:
this.$store.state.userCollection.userABC._.data.userName

// this makes people fully aware that they actually work with modules,
// not just a dumb plain object
this.$storeX['users/abc'].state.data.userName
this.$storeX['users/abc'].state.metadata.createdBy
this.$storeX['users/abc'].state.local.loginTime
this.$storeX['users/abc/tasks/xyz'].state.data.taskName
// we can create a clean, dedicated `modules` object for the user to iterate on
this.$storeX['users/abc/tasks/xyz'] === this.$storeX['users/abc/tasks'].modules.xyz  // true
this.$storeX['users/abc/tasks'].xyz // undefined

// works as a reactive object in templates
{{ $storeX['users/abc'].state.data.userName }}

// potentially, we dispatch to the module directly
const module = this.$storeX['users/abc']
module.dispatch('changeName', 'Lucas')

// although the classic Vuex dispatch syntax might also be kept for
// those who are more familiar with it
this.$storeX.dispatch('users/abc/changeName', 'Lucas')
// VueSync action
this.$storeX.dispatch('users/abc/set', { name: 'Lucas' })
// there is still sometimes a need for dispatching root actions
this.$storeX.dispatch('actionName', {})

// if there is a need to change the metadata, or work with local properties, specify it
// in the options
module.dispatch('set', { createdBy: 'IDlucas456xyz' }, { target: 'metadata' })
module.dispatch('set', { loginTime: 897654321 }, { target: 'local' })
```
